src/basic/errno-util.h

```diff
@@ -73,6 +73,16 @@ static inline int RET_NERRNO(int ret) {
         return ret;
 }
 
+/* Collect possible errors in <acc>, so that the first error can be returned.
+ * Returns (possibly updated) <acc>. */
+#define RET_GATHER(acc, err)                    \
+        ({                                      \
+                int *__a = &(acc), __e = (err); \
+                if (*__a >= 0 && __e < 0)       \
+                        *__a = __e;             \
+                *__a;                           \
+        })
```

Comment:
Interesting. I like this form if it works.
Maybe using `typeof()` is better? For the case e.g. `ssize_t`.
```
#define RET_GATHER(r) \
        *((r) < 0 ? &(typeof(r)) { 0 } : &r)
```

Reply:
This is some weird ass syntax. And I had to wrap my brain around three times before I could grok it. I don't like it as it raises the cognitive barrier for little gain. I think this is even the first time I see a macro used on the left side of an assignment.

Reply:
There are two related patterns: one where we just collect the first negative error, and the second pattern where we collect the first negative error, but also do something with the positive values. The second pattern is actually more of a class of patterns, because the details of how the positive values are handled vary. It turns out that the first pattern is much more common, and this is what the `RET_GATHER` macro currently implements. The proposed pattern would implement the second.

So maybe it can be done, but we'd at least need to go over all the callsites and verify that returning positive is handled the same as returning 0.

Reply:
```diff
diff --git src/basic/errno-util.h src/basic/errno-util.h
index f477db7852..dd7c8e93e2 100644
--- src/basic/errno-util.h
+++ src/basic/errno-util.h
@@ -83,6 +83,12 @@ static inline int RET_NERRNO(int ret) {
                 *__a;                           \
         })
 
+#define RET_GATHER2(acc)                        \
+        *((acc) < 0 ? &(int) { 0 } : &acc)
+
+// *((acc) < 0 ? &(typeof(acc)) { 0 } : &r)
+
+
 static inline int errno_or_else(int fallback) {
         /* To be used when invoking library calls where errno handling is not defined clearly: we return
          * errno if it is set, and the specified error otherwise. The idea is that the caller initializes
diff --git src/test/test-errno-util.c src/test/test-errno-util.c
index 62a508c4c9..9fde4960da 100644
--- src/test/test-errno-util.c
+++ src/test/test-errno-util.c
@@ -89,4 +89,19 @@ TEST(RET_GATHER) {
         assert_se(y == 3);
 }
 
+TEST(RET_GATHER2) {
+        int x = 0, y = 2;
+
+        RET_GATHER2(x) = 5;
+        assert_se(x == 5);
+        RET_GATHER2(x) = -5;
+        assert_se(x == -5);
+        RET_GATHER2(x) = -1;
+        assert_se(x == -5);
+
+        RET_GATHER2(x) = y++;
+        assert_se(x == -5);
+        assert_se(y == 3);
+}
+
 DEFINE_TEST_MAIN(LOG_INFO);
```